dbft_test.go

```diff
@@ -451,6 +451,230 @@ func TestDBFT_Invalid(t *testing.T) {
+
+	// Step 7. The backup (at view 0) replica 3 decides to change its view
+	// (possible on timeout) and sends the ChangeView message.
+	s3.OnReceive(resp0V0)
+	s3.OnReceive(resp2V0)
+	s3.OnTimeout(timer.HV{Height: r3.currHeight + 1, View: 0})
+	cv3V0 := r3.tryRecv()
+	require.NotNil(t, cv3V0)
+	require.Equal(t, payload.ChangeViewType, cv3V0.Type())
+
+	// Step 2. The primary (at view 0) replica 1 decides to change its view
+	// (possible on timeout after receiving at least M non-commit messages from the
+	// current view) and sends the ChangeView message.
+	s1.OnReceive(resp0V0)
+	s1.OnReceive(cv3V0)
+	s1.OnTimeout(timer.HV{Height: r1.currHeight + 1, View: 0})
+	cv1V0 := r1.tryRecv()
+	require.NotNil(t, cv1V0)
+	require.Equal(t, payload.ChangeViewType, cv1V0.Type())
+
+	// Step 4. The backup (at view 0) replica 0 decides to change its view
+	// (possible on timeout after receiving at least M non-commit messages from the
+	// current view) and sends the ChangeView message.
+	s0.OnReceive(cv3V0)
+	s0.OnTimeout(timer.HV{Height: r0.currHeight + 1, View: 0})
+	cv0V0 := r0.tryRecv()
+	require.NotNil(t, cv0V0)
+	require.Equal(t, payload.ChangeViewType, cv0V0.Type())
+
+	// Step 8. The primary (at view 0) replica 1 collects M ChangeView messages
+	// (from itself and replicas 1, 3) and changes its view to 1.
+	s1.OnReceive(cv0V0)
+	require.Equal(t, uint8(1), s1.ViewNumber)
+
+	// Step 9. The backup (at view 0) replica 0 collects M ChangeView messages
+	// (from itself and replicas 0, 3) and changes its view to 1.
+	s0.OnReceive(cv1V0)
+	require.Equal(t, uint8(1), s0.ViewNumber)
+
+	// Step 10. The primary (at view 1) replica 0 sends the PrepareRequest message.
+	s0.OnTimeout(timer.HV{Height: r0.currHeight + 1, View: 1})
+	reqV1 := r0.tryRecv()
+	require.NotNil(t, reqV1)
+	require.Equal(t, payload.PrepareRequestType, reqV1.Type())
+
+	// Step 11. The backup (at view 1) replica 1 receives the PrepareRequest of
+	// view 1 and sends the PrepareResponse.
+	s1.OnReceive(reqV1)
+	resp1V1 := r1.tryRecv()
+	require.NotNil(t, resp1V1)
+	require.Equal(t, payload.PrepareResponseType, resp1V1.Type())
+
+	// Steps 12, 13 will be performed later, see the comments to Step 12, 13.
+
+	// Step 14. The backup (at view 0) replica 3 collects M ChangeView messages
+	// (from itself and replicas 0, 1) and changes its view to 1.
+	s3.OnReceive(cv0V0)
+	s3.OnReceive(cv1V0)
+	require.Equal(t, uint8(1), s3.ViewNumber)
+
+	// Intermediate step A. It is added to make step 14 possible. The backup (at
+	// view 1) replica 3 doesn't receive anything for a long time and sends
+	// RecoveryRequest.
+	//
+	// IMPORTANT NOTE (possibly, a bug in dBFT): probably, we shouldn't take into
+	// account Recovery messages on checking whether we can perform the CV. Doing
+	// so will allow to avoid the current deadlock situation, because replicas
+	// 0 and 1 can't change view without knowing anything from the replica 3, see
+	// the comment to step 12.
```

Comment:
This is the escond thing I'd like to discuss. Node can't send the `ChangeView` on timeout if it hasn't received messages from more than F nodes (it will send the `RecoveryRequest` instead and wait messages from at least M nodes). The problem is that *any* message will suite, even `RecoveryRequest`. That's why the deadlock described in this test is possible. What I suggest is to count only "meaningful" messages from at least `M` nodes to make the desicion to perform `ChangeView`.

Reply:
Then it'll never be able to send CV. Suppose primary is dead, everyone else sends a recovery request. Everyone receives everything, but ignore and can't make a CV.

Reply:
You're right, I didn't consider this case.

Reply:
The comment is adjusted.